cdap-common/src/main/java/co/cask/cdap/common/kerberos/OwnerAdmin.java

```diff
@@ -0,0 +1,100 @@
+ */
+public interface OwnerAdmin {
+
+  /**
+   * Store the owner's kerberosPrincipalId for the given {@link EntityId}
+   *
+   * @param entityId The {@link EntityId} whose owner kerberosPrincipalId needs to be stored
+   * @param kerberosPrincipalId the {@link KerberosPrincipalId} of the {@link EntityId} owner
+   * @throws IOException if failed to get the store
+   * @throws AlreadyExistsException if the given entity already has an owner
+   * @throws IllegalArgumentException if the given KerberosPrincipalId is not valid or the entity is not of
+   * supported type.
+   */
+  @Beta
+  void add(NamespacedEntityId entityId, KerberosPrincipalId kerberosPrincipalId)
+    throws IOException, AlreadyExistsException;
+
+  /**
+   * Retrieves the owner information for the given {@link EntityId}
+   *
+   * @param entityId the {@link EntityId} whose owner principal information needs to be retrieved
+   * @return {@link KerberosPrincipalId} of the {@link EntityId} owner
+   * @throws IOException if failed to get the store
+   * @throws IllegalArgumentException if the given entity is not of supported type.
+   */
+  @Beta
+  @Nullable
+  KerberosPrincipalId getOwner(NamespacedEntityId entityId) throws IOException;
+
+  /**
+   * <p>
+   * Returns the {@link KerberosPrincipalId} to be impersonated for this {@link EntityId} by tracing the entity
+   * hierarchy.
+   * </p>
+   * <p>
+   * If an owner is present for this entity id then returns the {@link KerberosPrincipalId} of that immediate owner.
+   * If a direct owner is not present then the namespace owner {@link KerberosPrincipalId} will be returned if
+   * one is present else returns null.
+   * <p>
+   *
+   * @param entityId the {@link EntityId} whose owner principal information needs to be retrieved
+   * @return {@link KerberosPrincipalId} of the effective owner for the given entity
+   * @throws IOException if  failed to get the store
+   * @throws IllegalArgumentException if the given entity is not of supported type.
+   */
+  @Beta
+  @Nullable
+  KerberosPrincipalId getEffectiveOwner(NamespacedEntityId entityId) throws IOException;
+
+  /**
+   * Checks if owner information exists or not
+   *
+   * @param entityId the {@link EntityId} for which the check needs to be performed
+   * @return a boolean true: owner principal exists, false: no owner principal exists
+   * @throws IOException if failed to get the store
+   * @throws IllegalArgumentException if the given entity is not of supported type.
+   */
+  @Beta
+  boolean exists(NamespacedEntityId entityId) throws IOException;
+
+
+  /**
+   * Deletes the owner principal for the given {@link EntityId}
+   *
+   * @param entityId the entity whose owner principal needs to be deleted
+   * @throws IOException if failed to get the owner store
+   * @throws IllegalArgumentException if the given entity is not of supported type.
+   */
+  @Beta
+  void delete(NamespacedEntityId entityId) throws IOException;
```

Comment:
We choose to make it this way so that the delete can be performed on entities which have owner or not. Else we will need to first check for existence in store and then do delete causing us to add if block at various places. Is this bad ?